client/consensus/grandpa/authorities.go

```diff
@@ -160,51 +231,414 @@ func (authSet *AuthoritySet) addForcedChange(pending PendingChange, isDescendent
+	}
+	return applied, nil
 }
 
-// InnerLocked
-// Returns access to the [`AuthoritySet`] and locks it.
+// PendingChange A pending change to the authority set.
 //
-// For more information see [`SharedDataLocked`].
-// TODO update with shared logic
-func (sas *SharedAuthoritySet) InnerLocked() AuthoritySet {
-	return sas.authoritySet
+// This will be applied when the announcing block is at some depth within
+// the finalized or unfinalized chain.
+type PendingChange struct {
+	// The new authorities and weights to apply.
+	nextAuthorities AuthorityList
+	// How deep in the chain the announcing block must be
+	// before the change is applied.
+	delay uint
+	// The announcing block's height.
+	canonHeight uint
+	// The announcing block's hash.
+	canonHash common.Hash
+	// The delay kind.
+	delayKind DelayKind
+}
+
+// EffectiveNumber Returns the effective number this change will be applied at.
+func (pc *PendingChange) EffectiveNumber() uint {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced change.
+type AuthoritySetChanges []authorityChange
+
+// append an authorityChange to AuthoritySetChanges
+func (authSetChanges *AuthoritySetChanges) append(setId uint64, blockNumber uint) {
+	*authSetChanges = append(*authSetChanges, authorityChange{
+		setId:       setId,
+		blockNumber: blockNumber,
+	})
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+// Latest => bool
+// Set => &AuthorityChange
+// Unknown => nil
+// TODO for reviewers, this can be a VDT but I'm not sure its needed
+func (authSetChanges *AuthoritySetChanges) getSetId(blockNumber uint) (latest bool, set *authorityChange, err error) {
+	if authSetChanges == nil {
+		return false, nil, fmt.Errorf("getSetId: authSetChanges is nil")
+	}
+	authSet := *authSetChanges
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		return true, nil, nil // Latest case
+	}
+
+	idx, _ := searchSetChanges(blockNumber, authSet)
+	if idx < len(authSet) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setId != 0 {
+			return false, nil, nil // Unknown case
+		}
+
+		return false, &authChange, nil // Set case
+	}
+
+	return false, nil, nil // Unknown case
+}
+
+func (authSetChanges *AuthoritySetChanges) insert(blockNumber uint) {
+	var idx int
+	if authSetChanges == nil {
+		idx = 0
+	} else {
+		idx, _ = searchSetChanges(blockNumber, *authSetChanges)
+	}
+
+	set := *authSetChanges
+
+	var setId uint64
+	if idx == 0 {
+		setId = 0
+	} else {
+		setId = set[idx-1].setId + 1
+	}
+
+	// TODO reviewer double check this logic
+	if !(idx == len(set) || set[idx].setId != setId) {
```

Comment:
In the current way is quite hard to understand the condition (I get lost 🤣 ), if you want you can apply the [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) and you will have:

```go
	if idx != len(set) && set[idx].setId == setId {
```

For example, if we have a set of 3 items and the idx is `4` it will panic since it will face an out-of-bounds error, I would suggest changing the first condition to be a guard `if idx < len(set) && set[idx].setId == setId` and change the idx to `uint` type then you don't have any problems with negative numbers.

So, it is not possible to have more than one change per set id, right? If so instead of panicking we can have an error `ErrChangeForSetIDAlreadyExists` and return it

Reply:
The rust code is `assert!(idx == self.0.len() || self.0[idx].0 != set_id);` which makes me think we should panic right? I think this is probably gets enforced that we cannot have more than one change per set ID right? My gut is to panic since they do, but also if you don't think its best I am more than happy to change it.

Your suggested change to use a `<` sign only applies if we error instead of panic right?